### PR TITLE
[WIP] Channels last refactor

### PR DESCRIPTION
**Pull request by jjsjann123**
Channels Last support in nvfuser

Background:
To support channels last in nvfuser with optimal performance, we want to allow dimension collapsing in generated code on channels-last tensors, which greatly simplifies indexing.
Current API in codegen only allows dimensional collapsing on neighboring axes. The unfortunate thing is that memory format design in PyTorch is implicitly marked by strides, while the semantics meaning of axes remain unchanged. i.e. A 4d tensor with axes [N, C, H, W] would have the same shape in both format, while contiguous tensor carries strides [C*H*W, H*W, W, 1] and channels-last tensor [H*W*C, 1, W*C, C].

Approach:
We identify input tensor in channels-last format and permute them to NHWC. This creates an inconsistency between codegen tensor and TorchScript tensor. Our parser handles and propagates memory format accordingly. I.e., consumes and produces channels-last inputs when it can, while transposes inputs to original format and output non-permuted outputs.
Fusion inputs/outputs in channels-last format is marked and permuted before/after fusion execution to ensure correctness on the interfacing between nvfuser and TorchScript.

- [x] add simple cpp test to ensure simplified indexing in generated code.
~add python tests to verify nhwc fp16 inputs is handled properly.~ It has been handled in recent bfloat PR

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp
Comment (kevinstephano) [inline torch/csrc/jit/codegen/cuda/kernel_cache.cpp]: Nit: maybe the `inputs_` parameters should be left as `inputs` and the local variable be changed to `perm_inputs` so as not to confuse the input parameter as a data member?

**Comment by jjsjann123** [comment]
bump threads for review :pray:

**Comment by naoyam** [comment]
> Current API in codegen only allows dimensional collapsing on neighboring axes.

I think we can change this. Can you please file an issue?

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/kernel_cache.cpp:155]: > we can permute input tensor to have axes in decending order by their strides,

This makes sense. Do we need to make it specific to channels last? This seems like a workaround that could be generally useful for tensors whose axes are not ordered in the descending way.

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/kernel_cache.cpp:155]: That's true. Arbitrarily permuted tensor could be handled in similar ways and it would be beneficial to support them especially for memory-format-agnostic operations like pointwise ops. And for ops like batch_norm, where semantics of axes matters, we could just permute the tensor to supported memory-format.

I'm a little bit reluctant to do that right now as: 1. there's not a real use-case that those arbitrarily permuted tensor is relevant; 2. pytorch API only natively/actively supports contiguous & channels_last.

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/kernel_cache.cpp:155]: Sounds reasonable as a start. We could make it more generic as needed.

File: torch/csrc/jit/codegen/cuda/parser.cpp
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/parser.cpp]: Just wanted to quickly confirm if I'm understanding correctly. In these multi-input ops if `c10::nullopt` is passed as format, the output will be c_last if any one of the input is c_last, would that be correct?

File: torch/csrc/jit/codegen/cuda/parser.cpp
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/parser.cpp:146]: nit: I agree that deleting this constructor would be nicer.

File: torch/csrc/jit/codegen/cuda/parser.cpp
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/parser.cpp:111]: nit: re-computed hash.

File: torch/csrc/jit/codegen/cuda/parser.cpp
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/parser.cpp]: nit: error message.

File: torch/csrc/jit/codegen/cuda/parser.cpp
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/parser.cpp:203]: Please let me know if segmenter has any problem with these IO transpose ops. I don't think it will have correctness issue as is but could need updating.

File: torch/csrc/jit/codegen/cuda/parser.cpp
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/parser.cpp]: Also maybe I missed this part but would we be able to handle implicit broadcast on channel dimension in this case ? e.g. 
```
 A[N,C,H,W] + B[H,W]
```

**Review by shmsong — COMMENTED** [review]
Overall looks good to me. Had some question on memory format promotion rule.

File: torch/csrc/jit/codegen/cuda/parser.cpp
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp:203]: That was a thing I struggled earlier. But per Christian's suggestion, we now move the permutation logic onto the vanilla `Fusion`, which is not segmented. So we should be good there.

File: torch/csrc/jit/codegen/cuda/parser.cpp
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp]: Good catch!!! I don't think it works, I need to update python test to cover that.

As a quick WAR, we could be disabling format propagation for implicit broadcast and permute all inputs to contiguous.
Alternatively, we can unsqueeze contiguous tensors, so they would have a broadcast dim towards the end... I need to go over the list of ops to double check if there's any exceptions.

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp]: >Just wanted to quickly confirm if I'm understanding correctly. In these multi-input ops if c10::nullopt is passed as format, the output will be c_last if any one of the input is c_last, would that be correct?

And yes this is correct

File: torch/csrc/jit/codegen/cuda/fusion.h
Comment (kevinstephano) [inline torch/csrc/jit/codegen/cuda/fusion.h]: I talked to you about this live.  The naming here seems confusing to me where you are marking an Input as Channels last but the mechanism by which you do so is recording the index of the specific input.  Perhaps it would be less confusing if you just named the method `setChannelsLastInput()` and remove the fact that you are specifying the input via an index from the name.

**Review by kevinstephano — COMMENTED** [review]
I added a couple of comments.  The rest looked okay to me.

**Comment by jjsjann123** [comment]
Updated per review comments. bumping to get a second review/stamp :bow:

**Review by kevinstephano — APPROVED** [review]
Adding my approval.

**Pull request merged**
